src/libponyrt/actor/actor.c

```diff
@@ -160,21 +166,31 @@ bool ponyint_actor_run(pony_ctx_t* ctx, pony_actor_t* actor, size_t batch)
       // If we handle an application message, try to gc.
       app++;
       try_gc(ctx, actor);
-
-      if(app == batch)
-        return !has_flag(actor, FLAG_UNSCHEDULED);
     }
 
+    msgs++;
+
     // Stop handling a batch if we reach the head we found when we were
-    // scheduled.
-    if(msg == head)
+    // scheduled or if we've hit our batch size
+    if(msgs == batch || msg == head)
       break;
   }
 
+  // Make sure we gc at least once
+  try_gc(ctx, actor);
```

Comment:
I'm not really thrilled with the fact that if the last message we handled was an application message we do an extra gc run here. Thoughts on how to handle? I'm not sure as at Sendence we've changed to only attempt a GC after each batch is done.